app/models/manageiq/providers/inflector.rb

```diff
@@ -0,0 +1,17 @@
+module ManageIQ::Providers::Inflector
+  class ObjectNotNamespacedError < StandardError; end
+
+  def self.provider_name(class_or_instance)
+    klass = class_or_instance.class == Class ? class_or_instance : class_or_instance.class
+    provider_module(klass).name.split('::').last
+  end
+
+  def self.provider_module(klass, original_object = nil)
+    if klass == Object
+      raise ObjectNotNamespacedError, "Cannot get provider module from non namespaced object #{original_object}"
+    end
+
+    klass.parent == ManageIQ::Providers ? klass : provider_module(klass.parent, klass)
+  end
+  private_class_method :provider_module
```

Comment:
I revisited that code and like to keep the recursive version, because the splitting on the class name would still require checks for being in the right namespace and the name being not nil.
So I think this version is cleaner